source/slang/slang-emit-cpp.cpp

```diff
@@ -1695,11 +1703,22 @@ void CPPSourceEmitter::_emitWitnessTableWrappers()
                             if (reqParamType->op == kIROp_RawPointerType &&
                                 param->getFullType()->op != kIROp_RawPointerType)
                             {
-                                m_writer->emit("*static_cast<");
-                                emitType(param->getFullType());
-                                m_writer->emit("*>(");
-                                m_writer->emit(getName(param));
-                                m_writer->emit(")");
+                                if (as<IRPtrTypeBase>(param->getFullType()))
+                                {
```

Comment:
I don't think we need to do anything right now, but the thing I'd argued for is having a named predicate that tests for whatever property of a type we care about here:

```
bool doesTypeNeedAddedIndirectionWhenPassedGenerically(IRType* type);
```

Part of my point is that I don't actually know exactly what property of the types involved is being tested for here, and that means the code isn't really self-documenting right now.

The eventual goal should be to have a well-defined vocabulary for talking about types, and predicates that can test for whether types do or do not have the terms in that vocabulary.

Some examples of properties we might eventually need/want:

* A type can either have a fixed/known size, or not. Non-concrete generic, associated, and `This` types are all not fixed-size.

* A type can be "memory-only" meaning that in final generated code it cannot be stored used as the type of an SSA temporary. Types that are not fixed-size must be memory-only. Our compiler needs to ensure that we eliminate any SSA temporaries of memory-only types before emitting code, either by specializing them out or by applying you `alloca()`-based legalization.

* Two types can be layout-compatible if there is a bidirectional value-preserving conversion between the two that consists of just reinterpreting bits (e.g., a `T*` and a `RawPtr` are always layout-compatible).

I'm not saying those are the properties we nee because, again, I'm not 100% clear on the predicate we are testing for here. The fact that the predicate needs to agree between callers and callees means that having a name for the properties and subroutines to check them seems important.